README.md
URL2PFD and Download URLS as PDFs [Version 6.2]
===============================================

URL2PDF is a commandline utility that downloads URLs as PDFs. Included is an action for Apple's Automator called Download URLS as PDFs, which allows the user to archive a web page as a PDF document in a given workflow.

Scott Garner
scott@j38.net
http://scott.j38.net/

forked by 
Robert Welz
welz.willi@gmail.com
http://rwelz.byethost13.com

#Change Log and release notes:
see file changelog.txt


Doku/changelog.txt
6/6/17 version 6.2.6 by Robert Welz (welz.willi@gmail.com)
- java script functions can be called in own service by entering all functions separated by quotes and commas, for example for https://www.macupdate.com/app/mac/18510/devonthink-pro you would enter
"javascript:$.showMore('description')","javascript:$.showMore('release_notes')"
in the advanced tab of service in Automator.


3/10/17 version 6.2.5 by Robert Welz (welz.willi@gmail.com)
- added "open folder" and "open file" options to open folder which contains pdf or open pdf in default viewer after download.

3/10/17 version 6.2.4 by Robert Welz (welz.willi@gmail.com)
Better example workflow

3/9/17 version 6.2.3 by Robert Welz (welz.willi@gmail.com)
- new project layout - now its only one project. Just build the "BuildAll" target.

3/9/17 version 6.2.2 by Robert Welz (welz.willi@gmail.com)

- xib can be edited without Xcode crashing.

1/28/17 version 6.2.1 by Robert Welz (welz.willi@gmail.com)

cli client “url2pdf”
- executing without options shows help page
- by default pdfs are stored in /Users/USERNAME/URL2PDF Output
which will be created if missing
- made compile with Xcode 8.2.1 without warnings


- written a INSTALL documentation in german and english with Pages and lots of screenshots. Exported to pdf, too.

The original project can be found at: https://github.com/scottgarner/URL2PDF
This version can be found at: http://rwelz.byethost13.com/?p=199

2013-06-12 Version 6.2

- Fix for paths with spaces
- Page size fixes

2012-05-30 Version 6.0

- Changed: The action is now a wrapper for a commandline utility.
- Changed: Lion compatibility.
- Chagned: Source now available on Github.
- Changed: Versioning system.

2011-02-22 Version 0.5.0

- Changed:  Snow Leopard compatibility.

2007-11-04 Version 0.4.2

- Fixed: Paginated export works with stand-alone app again.

2007-11-04 Version 0.4.1

- Fixed: Version number issues.
- Fixed: Filename from page title works again.

2007-11-03 Version 0.4.0

- Changed:  Leopard compatibility.
	
2006-04-19 Version 0.3.1

- Changed:  Universal Binary (Maybe...if I did it right.)

2005-07-28 Version 0.3

- New:  Donation button (because you know you want to).
- New:  Advanced option to enable/disable image loading.
- New:  Advanced option to enable/disable JavaScript.
- Fixed:  Pages with forward slashes in their title no longer cause an error when title-based printing is enabled.
- Fixed:  Better handling of pages with child frames (; frames--not necessarily just HTML frames).
- Changed:  Workaround for WebKit animated GIF bug.
- Changed:  Now released under the New BSD License.

2005-07-11 Version 0.2

- Initial "final" release.  Some bugs fixed and probably some new ones introduced.  Either way, the "BETA MODE" watermark is now history.

2005-06-30 Version 0.1 beta

- Initial beta release.

#Release Notes

I finally gave up on trying to update the original action to work in Lion and decided to take a completely different approach. All of the heavy lifting is now handled by a commandline utility called URL2PDF, which is wrapped by an AppleScript-based Automator Action. 

URL2PDF/url2pdf/PDFDownloader.m
//
//  PDFDownloader.m
//  DownloadPDF
//
//  Created by Scott Garner on 5/23/12.
//  Modified by welz.willi@gmail.com (c) 1/27/17
//  Copyright (c) 2012 Project J38. All rights reserved.
//

#import "PDFDownloader.h"

@implementation NSURLRequest (NSURLRequestWithIgnoreSSL)

+ (BOOL)allowsAnyHTTPSCertificateForHost:(NSString *)host
{
    return YES;
}

@end

@implementation PDFDownloader

@synthesize loadComplete;
@synthesize pageTitle;
@synthesize scripts;
@synthesize request;

- (id)downloadURLs:(id)input parameters: (NSMutableDictionary *) parameters
{
    // Retrieve Parameters
    
    NSString *savePath = [ parameters objectForKey:@"savePath"];
    int fileNameFrom = [[ parameters objectForKey:@"fileNameFrom"] intValue];
    
    int printOrientation = [[ parameters objectForKey:@"printOrientation"] intValue];
    BOOL printPaginate = [[ parameters objectForKey:@"printPaginate"] boolValue];
    BOOL printBackgrounds = [[ parameters objectForKey:@"printBackgrounds"] boolValue];	
    
    BOOL loadImages = [[ parameters objectForKey:@"loadImages"] boolValue];	
    BOOL enableJavaScript = [[ parameters objectForKey:@"enableJavaScript"] boolValue];	                                       
    
    self.scripts = [[ parameters objectForKey:@"scripts"] componentsSeparatedByString:@","];

    // Paper Size
    
    NSPrintInfo *printInfo = [NSPrintInfo sharedPrintInfo];
    NSSize pageSize = [printInfo paperSize];
    
	int printWidth;
	int printHeight;
	switch (printOrientation) {
		case 0:
			printWidth = pageSize.width;
			printHeight = pageSize.height;
			break;
		case 1:
			printWidth = pageSize.height;
			printHeight = pageSize.width;
			break;
	}	
    
    // Webview
    
    NSRect frame = NSMakeRect(0,0,1,1);
    
    WebView *webView = [[WebView alloc] initWithFrame:frame];
    [webView setMaintainsBackForwardList:NO];	
    [webView setFrameLoadDelegate:self];
    [webView setResourceLoadDelegate:self];
    [webView setMediaStyle:@"screen"];
    
    // Window
    
    NSWindow * window = [[NSWindow alloc]  
                         initWithContentRect:NSMakeRect(0,0,1024,768)                        
                         styleMask:NSBorderlessWindowMask                         
                         backing:NSBackingStoreNonretained defer:NO];
    [window setContentView:webView];    
    
    // Static Prefernces
    
    [[webView preferences] setAllowsAnimatedImages:NO];	
    [[webView preferences] setAllowsAnimatedImageLooping:NO];
    [[webView preferences] setPlugInsEnabled:NO];
    [[webView preferences] setJavaEnabled:NO];	
    [[webView preferences] setJavaScriptCanOpenWindowsAutomatically:NO];
    
    // Optional preferences
    
    [[webView preferences] setJavaScriptEnabled:enableJavaScript];
	[[webView preferences] setShouldPrintBackgrounds:printBackgrounds];
	[[webView preferences] setLoadsImagesAutomatically:loadImages];	        
    
	// Process each URL
    
    NSMutableArray *output = [NSMutableArray arrayWithCapacity:[input count]];
    NSEnumerator *enumerate = [input objectEnumerator];
    NSURL *curURL;	
	
	while (curURL = [enumerate nextObject]) {
        
        //NSLog(@"Downloading URL: %@", [curURL absoluteString]);
        
        // Send Requests
        
        bool isRunning;
		[self setPageTitle:nil];
		[self setLoadComplete:NO];  
        
        request = [NSURLRequest requestWithURL:curURL
                                                 cachePolicy:NSURLRequestUseProtocolCachePolicy
                                             timeoutInterval:15];
        
        [[webView mainFrame] loadRequest:request];
        
        // Loop while waiting for responses.
        
        NSDate* next = [NSDate dateWithTimeIntervalSinceNow:1.0]; 
        do {
            isRunning = [[NSRunLoop currentRunLoop] runMode:NSDefaultRunLoopMode beforeDate:next];
        } while (![self loadComplete]);
        
        [[webView mainFrame] stopLoading];
        
        // Filename Fuss
        
        NSString *saveFilePath = [self getFileNameAt:savePath from:fileNameFrom forURL:curURL];
        
        // Print It
        
        [self printWebView:webView fileName:saveFilePath paginate:printPaginate orientation:printOrientation];    
        
		[output addObject:saveFilePath];
        
	}
    
	return (output);    
}

#pragma mark Filename Handling

- (NSString *)getFileNameAt:(NSString *)savePath from:(int)fileNameFrom forURL:(NSURL *) url
{
    savePath = [savePath stringByExpandingTildeInPath];
    
    if ([self pageTitle] == nil)
        [self setPageTitle:@"Untitled"];
    
    NSString *saveFile;
    NSString *saveFilePath;		
    
    // Set filename source
    
    switch (fileNameFrom) {
        case 0:
            if ([[url path] length] > 1)
                saveFile = [[[[url path] lastPathComponent] stringByDeletingPathExtension]
                            stringByAppendingPathExtension:@"pdf"];
            else
                saveFile = [[url host] stringByAppendingPathExtension:@"pdf"];
            break;
        case 1:
            // No forward slashes are allowed in file names, so we replace them with a colon.
            saveFile = [[[[self pageTitle] componentsSeparatedByString:@"/"] componentsJoinedByString:@":"] stringByAppendingPathExtension:@"pdf"];
            break;
    }	
    
    saveFilePath = [savePath stringByAppendingPathComponent:saveFile];	
    
    // Don't overwrite existing files.
    
    int renameCounter=1;		
    while ([[NSFileManager defaultManager] fileExistsAtPath:saveFilePath]) {
        saveFilePath = [savePath stringByAppendingPathComponent:[NSString stringWithFormat:@"%@-%i.%@",[saveFile stringByDeletingPathExtension],renameCounter++,[saveFile pathExtension]]];
    }     
    
    // Return safe path
    
    return saveFilePath;
}

- (void)printWebView:(WebView *) webView fileName:(NSString *)filename paginate:(BOOL)printPaginate orientation:(int)printOrientation
{
	// Get Print View...
	NSView *printView = [[[webView mainFrame] frameView] documentView];
    
    [[[webView mainFrame] frameView] setAllowsScrolling:NO];
    
    if (printPaginate) {
		// To paginate we have to fake a print
		
		NSMutableDictionary *printInfoDict;
		printInfoDict = [NSMutableDictionary dictionaryWithDictionary:[[NSPrintInfo sharedPrintInfo] dictionary]];
		//[printInfoDict setObject:filename forKey:NSPrintSavePath];
        [printInfoDict setObject:[NSURL fileURLWithPath:filename] forKey:NSPrintJobSavingURL];
		
		NSPrintInfo *printInfo = [[NSPrintInfo alloc] initWithDictionary: printInfoDict];
		[printInfo setHorizontallyCentered:NO];
		[printInfo setVerticallyCentered:NO];
		[printInfo setJobDisposition:NSPrintSaveJob];		
		
		// Handle margins
        
		NSRect imageableBounds = [printInfo imageablePageBounds];
		NSSize paperSize = [printInfo paperSize];
		if (NSWidth(imageableBounds) > paperSize.width) {
			imageableBounds.origin.x = 0;
			imageableBounds.size.width = paperSize.width;
		}
		if (NSHeight(imageableBounds) > paperSize.height) {
			imageableBounds.origin.y = 0;
			imageableBounds.size.height = paperSize.height;
		}
        
		[printInfo setBottomMargin:NSMinY(imageableBounds)];
		[printInfo setTopMargin:paperSize.height - NSMinY(imageableBounds) - NSHeight(imageableBounds)];
		[printInfo setLeftMargin:NSMinX(imageableBounds)];
		[printInfo setRightMargin:paperSize.width - NSMinX(imageableBounds) - NSWidth(imageableBounds)];
        
		// Set orientation
		
		switch (printOrientation) {
			case 0:
                [printInfo setOrientation:NSPaperOrientationPortrait];
				break;
			case 1:
                [printInfo setOrientation:NSPaperOrientationLandscape];
				break;
		}			
        
		// Create print operation
		
		NSPrintOperation *printOp;
		printOp = [NSPrintOperation printOperationWithView:printView printInfo:printInfo];
		[printOp setShowsPrintPanel:NO];
		[printOp setShowsProgressPanel:NO];
		[printOp runOperation];
		        
    } else {
        // No pagination
        
        NSRect printRect = [printView frame];
        NSData *printData = [printView dataWithPDFInsideRect:printRect];
        [printData writeToFile:filename atomically:YES];    
    }
    
    printf("%s\n", [filename UTF8String]);
    
}

#pragma mark Webview Delegates

- (void)webView:(WebView*)sender didFailLoadWithError:(NSError *)error forFrame:(WebFrame *)frame
{
    NSLog(@"didFailLoadWithError");
}


- (void)webView:(WebView*)sender didStartProvisionalLoadForFrame:(WebFrame*)frame
{	
    if ([sender mainFrame] == frame) {
	//NSLog(@"didStartProvisionalLoadForFrame");
    }
}

- (void)webView:(WebView *)sender didFailProvisionalLoadWithError:(NSError *)error forFrame:(WebFrame *)frame
{
    printf("Failed to load URL\n Error: %s\n", [[error description] UTF8String]);
    exit(EXIT_FAILURE);    
    
}

- (void)webView:(WebView*)sender didCommitLoadForFrame:(WebFrame*)frame
{	
	if ([sender mainFrame] == frame) {
		//NSLog(@"didCommitLoadForFrame");
	}
}

- (void)webView:(WebView *)sender didReceiveTitle:(NSString *)title forFrame:(WebFrame *)frame
{
    
	if ([sender mainFrame] == frame)
    {
		[self setPageTitle:title];
	}
}

- (void)webView:(WebView*)sender didFinishLoadForFrame:(WebFrame*)frame
{
	if ([sender mainFrame] == frame)
    {
        if((self.scripts != nil) && ([self.scripts count] > 0))
        //if(self.scripts != nil)
        {
            for(NSString *script in self.scripts)
            {
                [[sender windowScriptObject] evaluateWebScript:script];
            }
        }
        [self setLoadComplete:YES];
	}

}

@end


URL2PDF/url2pdf/main.m
//
//  main.m
//  URL2PDF
//
//  Created by Scott Garner on 5/29/12.
//  Modified by welz.willi@gmail.com (c) 1/27/17
//  Copyright (c) 2012 Project J38. All rights reserved.
//

#import <Foundation/Foundation.h>
#import "WebKit/WebKit.h"
#import "PDFDownloader.h"
#include <getopt.h>

NSString * const kDefaultDirectory = @"~/URL2PDF Output/";

void printUsage() {
    printf("The original URL2PDF 6.1 (c) 2012 Scott Garner\n");
    printf("URL2PDF 6.2.6 (c) 2017 Robert Welz\n");
    printf("------------------------------------------------------\n");
    printf("Options:\n");
    printf("  --help                        -h      Displays this message\n");
    printf("  --url=<URL>                   -u      URL to download - needs to start with http:// or https://\n");
    printf("  --enable-javascript=<BOOL>    -j      Enable javascript, YES or NO - NO is default if paramter not given\n");
    printf("  --script=<STRING>             -s      Script function in web page to execute.\n                                        Needs --enable-javascript=YES or -j YES\n                                        For example: -s \"javascript:$.showMore('description')\",\"javascript:$.showMore('release_notes')\"\n                                        Multiple script functions must be separated by ',' but no whitespace.\n");
    printf("  --print-paginate=<BOOL>       -g      Enable pagination, YES or NO - NO is default if paramter not given\n");
    printf("  --print-backgrounds=<BOOL>    -b      Print Backgrounds, YES or NO - YES is default if paramter not given\n");
    printf("  --load-images=<BOOL>          -i      Load Images, YES or NO - YES is default if paramter not given\n");
    printf("  --print-orientation=<VALUE>   -o      Orientation, Portrait or Landscape - Landscape is default if paramter not given\n");
    printf("  --autosave-name=<VALUE>       -n      Filename source: URL or Title - Filename from page title is default if paramter not given\n");
    printf("  --autosave-path=<PATH>        -p      Save path - ~/URL2PDF Output/ is default if paramter not given\n");
    printf("  --open-file=<BOOL>            -c      Opens folder containing pdf after successful download\n");
    printf("  --open-folder=<BOOL>          -d      Opens pdf in viewer after successful download\n");
    printf("  --autosave-path=<PATH>        -p      Save path - ~/URL2PDF Output/ is default if paramter not given\n");
    printf("\n  Example:\n  url2pdf -u http://www.apple.de\n");
}

NSMutableDictionary* parseOptions(const int argc, char **argv) {
    int option;
    
    if (argc == 0) {
        printUsage();
        exit(EXIT_FAILURE);
    }
    
    // Defaults
    
    
    NSMutableDictionary *parameters = [[NSMutableDictionary alloc] initWithObjectsAndKeys:
                                       [NSNull null] , @"url",
                                       //kDefaultDirectory, @"savePath",
                                       [NSNumber numberWithInt:1], @"fileNameFrom",
                                       [NSNumber numberWithInt:1], @"printOrientation",
                                       [NSNumber numberWithBool:NO], @"printPaginate",
                                       [NSNumber numberWithBool:YES], @"printBackgrounds",
                                       [NSNumber numberWithBool:YES], @"loadImages",
                                       [NSNumber numberWithBool:NO], @"enableJavaScript",
                                       [NSNumber numberWithBool:NO], @"openFile",
                                       [NSNumber numberWithBool:NO], @"openFolder",
                                       nil];
    
    // Option Table
    
    char *shortOptions = "hu:j:s:g:b:i:o:n:c:d:p:";
    const struct option longOptions[] = {
        {"help",                no_argument,        NULL,   'h'},
        {"url",                 required_argument,  NULL,   'u'},
        {"enable-javascript",   required_argument,  NULL,   'j'},
        {"script",              required_argument,  NULL,   's'},
        {"print-paginate",      required_argument,  NULL,   'g'},
        {"print-backgrounds",   required_argument,  NULL,   'b'},
        {"load-images",         required_argument,  NULL,   'i'},
        {"print-orientation",   required_argument,  NULL,   'o'},
        {"autosave-name",       required_argument,  NULL,   'n'},
        {"autosave-path",       required_argument,  NULL,   'p'},
        {"open-file",           required_argument,  NULL,   'c'},
        {"open-folder",         required_argument,  NULL,   'd'},
        {NULL,                  0,                  NULL,   0},
    };
    
    bool noParameterGiven = YES;
    while ((option = getopt_long(argc, argv, shortOptions, longOptions, NULL)) != -1) {
        
        switch(option) {
            case 'h':
                noParameterGiven = NO;
                printUsage();
                exit(EXIT_SUCCESS);
                
            case 'f':
                noParameterGiven = NO;
                [parameters setObject:[NSString stringWithFormat:@"%s",optarg] forKey:@"pdf"];
                break;
                
            case 'u':
                noParameterGiven = NO;
                [parameters setObject:[NSString stringWithFormat:@"%s",optarg] forKey:@"url"];
                break;
                
            case 'j':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"enableJavaScript"];
                else if (strcasecmp(optarg,"NO") == 0)
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"enableJavaScript"];
                else {
                    printf("Invalid argument for --enable-javascript\n");
                    exit(EXIT_FAILURE);
                }
                break;
            
            case 's':
                noParameterGiven = NO;
                [parameters setObject:[NSString stringWithFormat:@"%s",optarg] forKey:@"scripts"];
                break;
                
            case 'g':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"printPaginate"];
                else if (strcasecmp(optarg,"NO") == 0)
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"printPaginate"];
                else {
                    printf("Invalid argument for --print-paginate\n");
                    exit(EXIT_FAILURE);
                }
                break;
                
            case 'i':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"loadImages"];
                else if (strcasecmp(optarg,"NO") == 0)
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"loadImages"];
                else {
                    printf("Invalid argument for --load-images\n");
                    exit(EXIT_FAILURE);
                }
                break;
                
            case 'b':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"printBackgrounds"];
                else if (strcasecmp(optarg,"NO") == 0)
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"printBackgrounds"];
                else {
                    printf("Invalid argument for --print-backgrounds\n");
                    exit(EXIT_FAILURE);
                }
                break;
                
            case 'o':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"Portrait") == 0)
                    [parameters setObject:[NSNumber numberWithInt:0] forKey:@"printOrientation"];
                else if (strcasecmp(optarg,"Landscape") == 0)
                    [parameters setObject:[NSNumber numberWithInt:1] forKey:@"printOrientation"];
                else {
                    printf("Invalid argument for --print-orientation\n");
                    exit(EXIT_FAILURE);
                }
                break;
                
            case 'n':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"URL") == 0)
                    [parameters setObject:[NSNumber numberWithInt:0] forKey:@"fileNameFrom"];
                else if (strcasecmp(optarg,"Title") == 0)
                    [parameters setObject:[NSNumber numberWithInt:1] forKey:@"fileNameFrom"];
                else {
                    printf("Invalid argument for --autosave-name\n");
                    exit(EXIT_FAILURE);
                }
                break;
                
            case 'p':
                noParameterGiven = NO;
                [parameters setObject:[[NSString stringWithFormat:@"%s",optarg]stringByExpandingTildeInPath] forKey:@"savePath"];
                break;
                
            case 'c':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"openFile"];
                else
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"openFile"];
                break;
                
            case 'd':
                noParameterGiven = NO;
                if (strcasecmp(optarg,"YES") == 0)
                    [parameters setObject:[NSNumber numberWithBool:YES] forKey:@"openFolder"];
                else
                    [parameters setObject:[NSNumber numberWithBool:NO] forKey:@"openFolder"];
                break;
                
            default:
                printUsage();
                exit(EXIT_FAILURE);
        }
    }
    
    if(noParameterGiven == YES)
    {
        printUsage();
        exit(EXIT_SUCCESS);
    }
    
    if ([parameters objectForKey:@"url"] == [NSNull null]) {
        printf("Missing required parameter --url\n");
        exit(EXIT_FAILURE);
    }
    
    return parameters;
}


int main(const int argc, char **argv)
{
    
    @autoreleasepool {
        
        [NSApplication sharedApplication];
        
        //        NSArray *input = [[NSArray alloc] initWithObjects:
        //                          [NSURL URLWithString:@"http://cargocollective.com/coryschmitz"],
        //                          [NSURL URLWithString:@"http://mareodomo.com/"],
        //                          [NSURL URLWithString:@"http://appleinsider.com/"],
        //                          [NSURL URLWithString:@"http://bing.com/"],
        //                          [NSURL URLWithString:@"http://google.com/"],
        //                          [NSURL URLWithString:@"http://yahoo.com/"],
        //                          nil];
        
        
        // parse argv for multiple repeating options and remove them from argv
        
        NSMutableArray *options = [NSMutableArray array];
        for (int i = 0; i < argc; i++) {
            NSString *str = [[NSString alloc] initWithCString:argv[i] encoding:NSUTF8StringEncoding];
            [options addObject:str];
        }
        NSLog(@"%@", [NSString stringWithFormat:@"received: %@", [options description]]);
        
        NSMutableArray *optionsFiltered = [NSMutableArray array];
        //NSMutableArray *scripts = [NSMutableArray array];
        NSString *scripts;
        
        for(int i = 0; i < [options count]; i++)
        {
            if([options[i] isEqualToString:@"--script"] || [options[i] isEqualToString:@"-s"])
            {
                if(options[i+1] != nil)
                {
                    scripts = options[i+1];
//                    [scripts addObject:options[i+1]];
                    i++;
                }
                
                
            }
            else
            {
                [optionsFiltered addObject:options[i]];
            }
        }

        int count = (int)[optionsFiltered count];
        char **argvec = (char **)malloc(sizeof(const char*)*count);
        for (int i = 0; i < count; i++)
        {
            argvec[i] = strdup([[optionsFiltered objectAtIndex:i] UTF8String]);
        }
        // end: parse argv for multiple repeating options and remove them from argv

        NSMutableDictionary *parameters = parseOptions(count, argvec);
        if(scripts != nil)
        {
            [parameters setObject:scripts forKey:@"scripts"];
        }

        BOOL openFolder = [[ parameters objectForKey:@"openFolder"] boolValue];
        BOOL openFile = [[ parameters objectForKey:@"openFile"] boolValue];
        NSString *savePath = [ parameters objectForKey:@"savePath"];
        
        if((savePath == nil) || ([savePath isEqualToString:@""]))
        {
            [parameters setObject:[NSString stringWithFormat:@"%s",[kDefaultDirectory UTF8String]] forKey:@"savePath"];
            
            // create directory
            
            NSFileManager *fileManager= [NSFileManager defaultManager];
            NSError *error = nil;
            if(![fileManager createDirectoryAtPath:[kDefaultDirectory stringByExpandingTildeInPath] withIntermediateDirectories:NO attributes:nil error:&error]) {
                // An error has occurred, do something to handle it
                NSError *undelyingError = [[error userInfo] objectForKey:NSUnderlyingErrorKey];
                if([undelyingError code] != 17) // suppress directory exist error
                    NSLog(@"Failed to create directory \"%@\". Error: %@", kDefaultDirectory, error);
            }
        }
        
        NSArray *input = [[NSArray alloc] initWithObjects:
                          [NSURL URLWithString:[parameters objectForKey:@"url"]],
                          nil];
                
        PDFDownloader *downloader = [[PDFDownloader alloc] init];
        
        NSArray *output = [downloader downloadURLs:input parameters:parameters];
        if(openFolder == YES)
        {
            if([output count] > 0)
            {
                NSURL *fileURL = [NSURL fileURLWithPath: output[0]];
                NSURL *folderURL = [fileURL URLByDeletingLastPathComponent];
                [[NSWorkspace sharedWorkspace] openURL: folderURL];
            }
        }
        if(openFile == YES)
        {
            for (NSString *savePath in output)
            {
                NSURL *fileURL = [NSURL fileURLWithPath: savePath];
                [[NSWorkspace sharedWorkspace] openURL: fileURL];
            }
        }
    }
    return 0;
}


